File: code/programs/typescript/unix-tools/tests/tar.test.ts
/**
 * Tests for tar -- an archiving utility.
 *
 * We test the exported business logic functions: createHeaderBlock,
 * parseHeaderBlock, createArchive, listArchive, extractArchive,
 * and stripPathComponents.
 */

import { describe, it, expect, beforeEach, afterEach } from "vitest";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import {
  createHeaderBlock,
  parseHeaderBlock,
  createArchive,
  listArchive,
  extractArchive,
  stripPathComponents,
  TarHeader,
} from "../src/tar.js";

// ---------------------------------------------------------------------------
// Helpers: temp directory management.
// ---------------------------------------------------------------------------

let tmpDir: string;

beforeEach(() => {
  tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "tar-test-"));
});

afterEach(() => {
  fs.rmSync(tmpDir, { recursive: true, force: true });
});

// ---------------------------------------------------------------------------
// createHeaderBlock / parseHeaderBlock: round-trip tests.
// ---------------------------------------------------------------------------

describe("createHeaderBlock", () => {
  it("should create a 512-byte block", () => {
    const header: TarHeader = {
      name: "test.txt",
      mode: 0o644,
      uid: 1000,
      gid: 1000,
      size: 100,
      mtime: 1700000000,
      typeflag: "0",
      linkname: "",
    };

    const block = createHeaderBlock(header);
    expect(block.length).toBe(512);
  });

  it("should contain the filename", () => {
    const header: TarHeader = {
      name: "hello.txt",
      mode: 0o644,
      uid: 0,
      gid: 0,
      size: 0,
      mtime: 0,
      typeflag: "0",
      linkname: "",
    };

    const block = createHeaderBlock(header);
    expect(block.toString("ascii", 0, 9)).toBe("hello.txt");
  });

  it("should contain the ustar magic string", () => {
    const header: TarHeader = {
      name: "test",
      mode: 0o755,
      uid: 0,
      gid: 0,
      size: 0,
      mtime: 0,
      typeflag: "0",
      linkname: "",
    };

    const block = createHeaderBlock(header);
    const magic = block.toString("ascii", 257, 262);
    expect(magic).toBe("ustar");
  });
});

describe("parseHeaderBlock", () => {
  it("should return null for all-zero block (end-of-archive)", () => {
    const block = Buffer.alloc(512, 0);
    expect(parseHeaderBlock(block)).toBeNull();
  });

  it("should round-trip a header correctly", () => {
    const original: TarHeader = {
      name: "test.txt",
      mode: 0o644,
      uid: 1000,
      gid: 100,
      size: 42,
      mtime: 1700000000,
      typeflag: "0",
      linkname: "",
    };

    const block = createHeaderBlock(original);
    const parsed = parseHeaderBlock(block);

    expect(parsed).not.toBeNull();
    expect(parsed!.name).toBe("test.txt");
    expect(parsed!.mode).toBe(0o644);
    expect(parsed!.uid).toBe(1000);
    expect(parsed!.gid).toBe(100);
    expect(parsed!.size).toBe(42);
    expect(parsed!.mtime).toBe(1700000000);
    expect(parsed!.typeflag).toBe("0");
  });

  it("should handle directory entries", () => {
    const header: TarHeader = {
      name: "subdir/",
      mode: 0o755,
      uid: 0,
      gid: 0,
      size: 0,
      mtime: 1700000000,
      typeflag: "5",
      linkname: "",
    };

    const block = createHeaderBlock(header);
    const parsed = parseHeaderBlock(block);

    expect(parsed!.name).toBe("subdir/");
    expect(parsed!.typeflag).toBe("5");
    expect(parsed!.size).toBe(0);
  });

  it("should handle long filenames (up to 100 chars)", () => {
    const longName = "a".repeat(99) + ".";
    const header: TarHeader = {
      name: longName,
      mode: 0o644,
      uid: 0,
      gid: 0,
      size: 0,
      mtime: 0,
      typeflag: "0",
      linkname: "",
    };

    const block = createHeaderBlock(header);
    const parsed = parseHeaderBlock(block);

    expect(parsed!.name).toBe(longName);
  });
});

// ---------------------------------------------------------------------------
// stripPathComponents: path manipulation.
// ---------------------------------------------------------------------------

describe("stripPathComponents", () => {
  it("should strip leading components", () => {
    expect(stripPathComponents("a/b/c/file.txt", 2)).toBe("c/file.txt");
  });

  it("should return empty string when stripping all components", () => {
    expect(stripPathComponents("a/b", 2)).toBe("");
  });

  it("should return name unchanged with count 0", () => {
    expect(stripPathComponents("a/b/c", 0)).toBe("a/b/c");
  });

  it("should strip one component", () => {
    expect(stripPathComponents("dir/file.txt", 1)).toBe("file.txt");
  });

  it("should handle names without separators", () => {
    expect(stripPathComponents("file.txt", 1)).toBe("");
  });

  it("should handle trailing slashes", () => {
    expect(stripPathComponents("a/b/c/", 1)).toBe("b/c");
  });

  it("should return empty for over-stripping", () => {
    expect(stripPathComponents("a", 5)).toBe("");
  });
});

// ---------------------------------------------------------------------------
// createArchive / listArchive: creating and listing archives.
// ---------------------------------------------------------------------------

describe("createArchive", () => {
  it("should create an archive with a single file", () => {
    fs.writeFileSync(path.join(tmpDir, "hello.txt"), "Hello, world!");

    const result = createArchive(["hello.txt"], tmpDir);

    expect(result.buffer.length).toBeGreaterThan(0);
    expect(result.entries).toContain("hello.txt");
  });

  it("should create an archive with multiple files", () => {
    fs.writeFileSync(path.join(tmpDir, "a.txt"), "AAA");
    fs.writeFileSync(path.join(tmpDir, "b.txt"), "BBB");

    const result = createArchive(["a.txt", "b.txt"], tmpDir);

    expect(result.entries).toContain("a.txt");
    expect(result.entries).toContain("b.txt");
  });

  it("should include directories recursively", () => {
    const subDir = path.join(tmpDir, "subdir");
    fs.mkdirSync(subDir);
    fs.writeFileSync(path.join(subDir, "nested.txt"), "nested content");

    const result = createArchive(["subdir"], tmpDir);

    expect(result.entries.some(e => e.includes("subdir"))).toBe(true);
    expect(result.entries.some(e => e.includes("nested.txt"))).toBe(true);
  });

  it("should handle empty files", () => {
    fs.writeFileSync(path.join(tmpDir, "empty.txt"), "");

    const result = createArchive(["empty.txt"], tmpDir);

    expect(result.entries).toContain("empty.txt");
  });

  it("should end with two zero blocks", () => {
    fs.writeFileSync(path.join(tmpDir, "test.txt"), "test");

    const result = createArchive(["test.txt"], tmpDir);
    const buf = result.buffer;

    // The last 1024 bytes should be all zeros.
    const lastTwo = buf.subarray(buf.length - 1024);
    let allZero = true;
    for (let i = 0; i < lastTwo.length; i++) {
      if (lastTwo[i] !== 0) { allZero = false; break; }
    }
    expect(allZero).toBe(true);
  });
});

describe("listArchive", () => {
  it("should list entries from an archive", () => {
    fs.writeFileSync(path.join(tmpDir, "a.txt"), "AAA");
    fs.writeFileSync(path.join(tmpDir, "b.txt"), "BBB");

    const archive = createArchive(["a.txt", "b.txt"], tmpDir);
    const entries = listArchive(archive.buffer);

    expect(entries).toHaveLength(2);
    expect(entries[0].name).toBe("a.txt");
    expect(entries[1].name).toBe("b.txt");
  });

  it("should list directories and their contents", () => {
    const subDir = path.join(tmpDir, "mydir");
    fs.mkdirSync(subDir);
    fs.writeFileSync(path.join(subDir, "file.txt"), "content");

    const archive = createArchive(["mydir"], tmpDir);
    const entries = listArchive(archive.buffer);

    expect(entries.some(e => e.name === "mydir/")).toBe(true);
    expect(entries.some(e => e.name === "mydir/file.txt")).toBe(true);
  });

  it("should report correct file sizes", () => {
    fs.writeFileSync(path.join(tmpDir, "sized.txt"), "12345");

    const archive = createArchive(["sized.txt"], tmpDir);
    const entries = listArchive(archive.buffer);

    expect(entries[0].size).toBe(5);
  });

  it("should handle empty archive (just zero blocks)", () => {
    const emptyArchive = Buffer.alloc(1024, 0);
    const entries = listArchive(emptyArchive);

    expect(entries).toHaveLength(0);
  });
});

// ---------------------------------------------------------------------------
// extractArchive: extracting from archives.
// ---------------------------------------------------------------------------

describe("extractArchive", () => {
  it("should extract a single file", () => {
    // Create archive.
    fs.writeFileSync(path.join(tmpDir, "hello.txt"), "Hello, world!");
    const archive = createArchive(["hello.txt"], tmpDir);

    // Extract to a different directory.
    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir);

    const extracted = extractArchive(archive.buffer, extractDir, {
      preservePermissions: false,
      keepOldFiles: false,
      stripComponents: 0,
      verbose: false,
      filterFiles: [],
    });

    expect(extracted).toContain("hello.txt");

    const content = fs.readFileSync(path.join(extractDir, "hello.txt"), "utf-8");
    expect(content).toBe("Hello, world!");
  });

  it("should extract directories", () => {
    const subDir = path.join(tmpDir, "mydir");
    fs.mkdirSync(subDir);
    fs.writeFileSync(path.join(subDir, "file.txt"), "content");

    const archive = createArchive(["mydir"], tmpDir);

    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir);

    extractArchive(archive.buffer, extractDir, {
      preservePermissions: false,
      keepOldFiles: false,
      stripComponents: 0,
      verbose: false,
      filterFiles: [],
    });

    expect(fs.existsSync(path.join(extractDir, "mydir"))).toBe(true);
    expect(fs.existsSync(path.join(extractDir, "mydir", "file.txt"))).toBe(true);
  });

  it("should respect keepOldFiles", () => {
    // Create archive.
    fs.writeFileSync(path.join(tmpDir, "file.txt"), "new content");
    const archive = createArchive(["file.txt"], tmpDir);

    // Create existing file in extract directory.
    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir);
    fs.writeFileSync(path.join(extractDir, "file.txt"), "old content");

    extractArchive(archive.buffer, extractDir, {
      preservePermissions: false,
      keepOldFiles: true,
      stripComponents: 0,
      verbose: false,
      filterFiles: [],
    });

    const content = fs.readFileSync(path.join(extractDir, "file.txt"), "utf-8");
    expect(content).toBe("old content");
  });

  it("should strip path components", () => {
    const nested = path.join(tmpDir, "a", "b");
    fs.mkdirSync(nested, { recursive: true });
    fs.writeFileSync(path.join(nested, "file.txt"), "deep content");

    const archive = createArchive(["a"], tmpDir);

    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir);

    extractArchive(archive.buffer, extractDir, {
      preservePermissions: false,
      keepOldFiles: false,
      stripComponents: 1,
      verbose: false,
      filterFiles: [],
    });

    // "a/b/file.txt" with strip 1 becomes "b/file.txt".
    expect(fs.existsSync(path.join(extractDir, "b", "file.txt"))).toBe(true);
  });

  it("should filter files during extraction", () => {
    fs.writeFileSync(path.join(tmpDir, "keep.txt"), "keep me");
    fs.writeFileSync(path.join(tmpDir, "skip.txt"), "skip me");

    const archive = createArchive(["keep.txt", "skip.txt"], tmpDir);

    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir);

    extractArchive(archive.buffer, extractDir, {
      preservePermissions: false,
      keepOldFiles: false,
      stripComponents: 0,
      verbose: false,
      filterFiles: ["keep.txt"],
    });

    expect(fs.existsSync(path.join(extractDir, "keep.txt"))).toBe(true);
    expect(fs.existsSync(path.join(extractDir, "skip.txt"))).toBe(false);
  });

  it("should handle multiple files", () => {
    fs.writeFileSync(path.join(tmpDir, "a.txt"), "AAA");
    fs.writeFileSync(path.join(tmpDir, "b.txt"), "BBB");
    fs.writeFileSync(path.join(tmpDir, "c.txt"), "CCC");

    const archive = createArchive(["a.txt", "b.txt", "c.txt"], tmpDir);

    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir);

    const extracted = extractArchive(archive.buffer, extractDir, {
      preservePermissions: false,
      keepOldFiles: false,
      stripComponents: 0,
      verbose: false,
      filterFiles: [],
    });

    expect(extracted).toHaveLength(3);
    expect(fs.readFileSync(path.join(extractDir, "a.txt"), "utf-8")).toBe("AAA");
    expect(fs.readFileSync(path.join(extractDir, "b.txt"), "utf-8")).toBe("BBB");
    expect(fs.readFileSync(path.join(extractDir, "c.txt"), "utf-8")).toBe("CCC");
  });

  it("should round-trip create + extract", () => {
    // Create files with various sizes.
    fs.writeFileSync(path.join(tmpDir, "small.txt"), "small");
    fs.writeFileSync(path.join(tmpDir, "medium.txt"), "x".repeat(1000));
    fs.writeFileSync(path.join(tmpDir, "empty.txt"), "");

    const archive = createArchive(["small.txt", "medium.txt", "empty.txt"], tmpDir);

    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir);

    extractArchive(archive.buffer, extractDir, {
      preservePermissions: false,
      keepOldFiles: false,
      stripComponents: 0,
      verbose: false,
      filterFiles: [],
    });

    expect(fs.readFileSync(path.join(extractDir, "small.txt"), "utf-8")).toBe("small");
    expect(fs.readFileSync(path.join(extractDir, "medium.txt"), "utf-8")).toBe("x".repeat(1000));
    expect(fs.readFileSync(path.join(extractDir, "empty.txt"), "utf-8")).toBe("");
  });
});

File: code/programs/rust/unix-tools/tests/tar_test.rs
//! # Integration Tests for tar
//!
//! These tests verify tar archive creation, listing, and extraction
//! using real files on the filesystem.

use unix_tools::tar_tool::*;
use std::fs;
use std::path::Path;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn temp_path(name: &str) -> String {
    std::env::temp_dir()
        .join(format!("tar_integ_{}", name))
        .to_string_lossy()
        .into_owned()
}

fn cleanup(path: &str) {
    let p = Path::new(path);
    if p.is_dir() {
        let _ = fs::remove_dir_all(p);
    } else {
        let _ = fs::remove_file(p);
    }
}

// ---------------------------------------------------------------------------
// Tests: Header encoding/decoding
// ---------------------------------------------------------------------------

#[cfg(test)]
mod header {
    use super::*;

    #[test]
    fn roundtrip() {
        let header = TarHeader {
            name: "hello.txt".to_string(),
            mode: 0o644,
            uid: 501,
            gid: 20,
            size: 13,
            mtime: 1700000000,
            typeflag: b'0',
            uname: "user".to_string(),
            gname: "staff".to_string(),
        };

        let bytes = header.to_bytes();
        let parsed = TarHeader::from_bytes(&bytes).unwrap();

        assert_eq!(parsed.name, "hello.txt");
        assert_eq!(parsed.mode, 0o644);
        assert_eq!(parsed.uid, 501);
        assert_eq!(parsed.gid, 20);
        assert_eq!(parsed.size, 13);
        assert_eq!(parsed.typeflag, b'0');
    }

    #[test]
    fn zero_block_is_eof() {
        let block = [0u8; 512];
        assert!(TarHeader::from_bytes(&block).is_none());
    }

    #[test]
    fn directory_typeflag() {
        let header = TarHeader {
            name: "mydir/".to_string(),
            mode: 0o755,
            uid: 0,
            gid: 0,
            size: 0,
            mtime: 0,
            typeflag: b'5',
            uname: String::new(),
            gname: String::new(),
        };

        let bytes = header.to_bytes();
        let parsed = TarHeader::from_bytes(&bytes).unwrap();
        assert_eq!(parsed.typeflag, b'5');
        assert!(parsed.name.contains("mydir"));
    }

    #[test]
    fn large_file_size() {
        let header = TarHeader {
            name: "big.bin".to_string(),
            mode: 0o644,
            uid: 0,
            gid: 0,
            size: 1048576, // 1 MB
            mtime: 0,
            typeflag: b'0',
            uname: String::new(),
            gname: String::new(),
        };

        let bytes = header.to_bytes();
        let parsed = TarHeader::from_bytes(&bytes).unwrap();
        assert_eq!(parsed.size, 1048576);
    }
}

// ---------------------------------------------------------------------------
// Tests: Create and list
// ---------------------------------------------------------------------------

#[cfg(test)]
mod create_list {
    use super::*;

    #[test]
    fn create_single_file() {
        let dir = temp_path("create_single");
        cleanup(&dir);
        fs::create_dir_all(&dir).unwrap();
        fs::write(format!("{}/test.txt", dir), "hello world").unwrap();

        let opts = TarOptions {
            directory: Some(dir.clone()),
            verbose: false,
        };
        let files = vec!["test.txt".to_string()];
        let (archive, _) = create_archive(&files, &opts).unwrap();

        let entries = list_archive(&archive, &TarOptions::default()).unwrap();
        assert_eq!(entries, vec!["test.txt"]);

        cleanup(&dir);
    }

    #[test]
    fn create_multiple_files() {
        let dir = temp_path("create_multi");
        cleanup(&dir);
        fs::create_dir_all(&dir).unwrap();
        fs::write(format!("{}/a.txt", dir), "aaa").unwrap();
        fs::write(format!("{}/b.txt", dir), "bbb").unwrap();
        fs::write(format!("{}/c.txt", dir), "ccc").unwrap();

        let opts = TarOptions {
            directory: Some(dir.clone()),
            verbose: false,
        };
        let files = vec!["a.txt".into(), "b.txt".into(), "c.txt".into()];
        let (archive, _) = create_archive(&files, &opts).unwrap();

        let entries = list_archive(&archive, &TarOptions::default()).unwrap();
        assert_eq!(entries.len(), 3);
        assert!(entries.contains(&"a.txt".to_string()));
        assert!(entries.contains(&"b.txt".to_string()));
        assert!(entries.contains(&"c.txt".to_string()));

        cleanup(&dir);
    }

    #[test]
    fn create_with_directory() {
        let dir = temp_path("create_dir");
        cleanup(&dir);
        fs::create_dir_all(format!("{}/subdir", dir)).unwrap();
        fs::write(format!("{}/subdir/inner.txt", dir), "inside").unwrap();

        let opts = TarOptions {
            directory: Some(dir.clone()),
            verbose: true,
        };
        let files = vec!["subdir".to_string()];
        let (archive, messages) = create_archive(&files, &opts).unwrap();

        assert!(!messages.is_empty());
        let entries = list_archive(&archive, &TarOptions::default()).unwrap();
        assert!(entries.iter().any(|e| e.contains("subdir")));

        cleanup(&dir);
    }

    #[test]
    fn verbose_list_shows_permissions() {
        let dir = temp_path("verbose_list");
        cleanup(&dir);
        fs::create_dir_all(&dir).unwrap();
        fs::write(format!("{}/file.txt", dir), "data").unwrap();

        let opts = TarOptions {
            directory: Some(dir.clone()),
            verbose: false,
        };
        let files = vec!["file.txt".to_string()];
        let (archive, _) = create_archive(&files, &opts).unwrap();

        let verbose_opts = TarOptions { verbose: true, ..Default::default() };
        let entries = list_archive(&archive, &verbose_opts).unwrap();
        assert!(entries[0].contains("rw"));

        cleanup(&dir);
    }

    #[test]
    fn empty_archive() {
        let archive = vec![0u8; 1024]; // Two zero blocks
        let entries = list_archive(&archive, &TarOptions::default()).unwrap();
        assert!(entries.is_empty());
    }
}

// ---------------------------------------------------------------------------
// Tests: Extract
// ---------------------------------------------------------------------------

#[cfg(test)]
mod extract {
    use super::*;

    #[test]
    fn extract_single_file() {
        let src = temp_path("extract_src");
        let dst = temp_path("extract_dst");
        cleanup(&src);
        cleanup(&dst);
        fs::create_dir_all(&src).unwrap();
        fs::write(format!("{}/data.txt", src), "extract me").unwrap();

        let opts = TarOptions {
            directory: Some(src.clone()),
            verbose: false,
        };
        let (archive, _) = create_archive(&["data.txt".into()], &opts).unwrap();

        fs::create_dir_all(&dst).unwrap();
        extract_archive(&archive, &dst, &TarOptions::default()).unwrap();

        let content = fs::read_to_string(format!("{}/data.txt", dst)).unwrap();
        assert_eq!(content, "extract me");

        cleanup(&src);
        cleanup(&dst);
    }

    #[test]
    fn extract_preserves_content() {
        let src = temp_path("preserve_src");
        let dst = temp_path("preserve_dst");
        cleanup(&src);
        cleanup(&dst);
        fs::create_dir_all(&src).unwrap();

        let content = "Line 1\nLine 2\nLine 3\n";
        fs::write(format!("{}/multi.txt", src), content).unwrap();

        let opts = TarOptions {
            directory: Some(src.clone()),
            verbose: false,
        };
        let (archive, _) = create_archive(&["multi.txt".into()], &opts).unwrap();

        fs::create_dir_all(&dst).unwrap();
        extract_archive(&archive, &dst, &TarOptions::default()).unwrap();

        let extracted = fs::read_to_string(format!("{}/multi.txt", dst)).unwrap();
        assert_eq!(extracted, content);

        cleanup(&src);
        cleanup(&dst);
    }

    #[test]
    fn extract_directory_structure() {
        let src = temp_path("dir_struct_src");
        let dst = temp_path("dir_struct_dst");
        cleanup(&src);
        cleanup(&dst);
        fs::create_dir_all(format!("{}/sub", src)).unwrap();
        fs::write(format!("{}/sub/nested.txt", src), "nested").unwrap();

        let opts = TarOptions {
            directory: Some(src.clone()),
            verbose: false,
        };
        let (archive, _) = create_archive(&["sub".into()], &opts).unwrap();

        fs::create_dir_all(&dst).unwrap();
        extract_archive(&archive, &dst, &TarOptions::default()).unwrap();

        assert!(Path::new(&format!("{}/sub", dst)).is_dir());
        let content = fs::read_to_string(format!("{}/sub/nested.txt", dst)).unwrap();
        assert_eq!(content, "nested");

        cleanup(&src);
        cleanup(&dst);
    }

    #[test]
    fn extract_verbose_reports_files() {
        let src = temp_path("verbose_extract_src");
        let dst = temp_path("verbose_extract_dst");
        cleanup(&src);
        cleanup(&dst);
        fs::create_dir_all(&src).unwrap();
        fs::write(format!("{}/v.txt", src), "verbose").unwrap();

        let opts = TarOptions {
            directory: Some(src.clone()),
            verbose: false,
        };
        let (archive, _) = create_archive(&["v.txt".into()], &opts).unwrap();

        fs::create_dir_all(&dst).unwrap();
        let verbose_opts = TarOptions { verbose: true, ..Default::default() };
        let messages = extract_archive(&archive, &dst, &verbose_opts).unwrap();
        assert!(messages.contains(&"v.txt".to_string()));

        cleanup(&src);
        cleanup(&dst);
    }
}

// ---------------------------------------------------------------------------
// Tests: Security
// ---------------------------------------------------------------------------

#[cfg(test)]
mod security {
    use super::*;

    #[test]
    fn rejects_path_traversal() {
        let header = TarHeader {
            name: "../../../etc/passwd".to_string(),
            mode: 0o644,
            uid: 0,
            gid: 0,
            size: 0,
            mtime: 0,
            typeflag: b'0',
            uname: String::new(),
            gname: String::new(),
        };

        let mut archive = Vec::new();
        archive.extend_from_slice(&header.to_bytes());
        archive.extend_from_slice(&[0u8; 1024]);

        let result = extract_archive(&archive, "/tmp/tar_integ_security", &TarOptions::default());
        assert!(result.is_err());
    }
}

// ---------------------------------------------------------------------------
// Tests: Error handling
// ---------------------------------------------------------------------------

#[cfg(test)]
mod errors {
    use super::*;

    #[test]
    fn create_nonexistent_file() {
        let opts = TarOptions::default();
        let result = create_archive(&["nonexistent_file_xyz.txt".into()], &opts);
        assert!(result.is_err());
    }
}
